API/SWODLR/product.py

```diff
@@ -23,7 +25,7 @@ def GetProductsOfCurrentUser(
         graphQlBody = '''
             {
             	currentUser {
-            		products {
+            		products(limit: ''' + f'{pageSize}, after: "{pageAfterId}"' + ''') {
```

Comment:
GraphQL variables mean utilizing placeholder values `$limit` and then stating the values for those placeholder values within the `variables` key in the request.

https://graphql.org/learn/queries/#variables

https://graphql.org/learn/serving-over-http/#post-request

These are some references for that behavior. For example, a request that utilizes GraphQL variables will look at this:
```
{
  "query": "query($limit: Int, $after: String) { currentUser { products(limit: $limit, after: $after) { id } } }",
  "variables: {
    "limit": 10,
    "after": null
  }
}
```

Reply:
And how will it make the code more readable or easier to understand? It just makes it more confusing as you need to jump around to understand what happens. It does not help for someone with low level of knowledge in GraphQL.

This line is straightforward. You have these input arguments and you put the value after them. Period, clear.
`products(limit: ''' + f'{pageSize}, after: "{pageAfterId}"' + ''') {`

In what you suggested you define another layer of variables. Putting the Python variable into another variable and than using that. Why do we need to define another 'limit' variable as Integer, when we already had that in python? It feels excessive and needlessly complicating the code.

In my opinion the one without that forced new variable layer are much more clean and easier to read. If we would use the limit multiple places in that call, it would make sens to make a variables so we need to update it at only one place, but for one argument adding that much extra code just makes it looks more complicated, harder to understand and creates maintenance overhead.
`{
  "query": "query($limit: Int, $after: String) { currentUser { products(limit: $limit, after: $after) { id } } }",
  "variables: {
    "limit": 10,
    "after": null
  }
}`
vs
`{
  "query": "{ currentUser { products(limit: 10, after: null) { id } } }",
}`

Having those variables does not decrease any logic from the python side, so it does not have any positive effect on the code.

Reply:
GraphQL variables remove the need to escape the input on the query side. It's the same logic that is utilized for SQL prepared statements - decoupling the query statement from the user input; removing the need to manually perform the string escaping/type casting logic.

The UI for example, utilizes GraphQL variables in its operation and it would be preferable to have these tests be unified with the way we expect users to use the API. This request was also made yesterday: https://github.com/podaac/swodlr-functional-tests/pull/5#discussion_r1488677045

Reply:
> GraphQL variables remove the need to escape the input on the query side.
You to still need to put the input into the same json, just a different part of it, which will look the same just adding a lot of variable related code that will make the query much more confusing.

> decoupling the query statement from the user input; removing the need to manually perform the string escaping/type casting logic.
It does not remove the need to do the logic. I still need to cast Boolean in Python to string as GraphQL can't handle 'False', but only 'false', which is not a valid parameter for Boolean in Python. Same with 'null', Python has 'None', and GraphQL unable to handle that. So no, adding that overhead won't remove any type casting logic/code on Python side.

So just to be clear, you want me to write overcomplicated code that has no practical usage over the simple one in the case of these API tests just because the GraphQL is capable of that? And you expect that everyone will use the complicated version instead of the simple one?

Reply:
I wouldn't go so far to say it's impractical; the graphql docs explain it is considered a 'best practice' to separate the dynamic components of the query from the static parts:

> It wouldn't be a good idea to pass these dynamic arguments directly in the query string, because then our client-side code would need to dynamically manipulate the query string at runtime, and serialize it into a GraphQL-specific format. Instead, GraphQL has a first-class way to factor dynamic values out of the query, and pass them as a separate dictionary. ... This is also in general a good practice for denoting which arguments in our query are expected to be dynamic - we should never be doing string interpolation to construct queries from user-supplied values.

Looking at how you've structured the code you would likely have to change your `GraphQlBase.Base_Post` function to accept an operation string and an optional variable dictionary

Reply:
> It wouldn't be a good idea to pass these dynamic arguments directly in the query string, because then our client-side code would need to dynamically manipulate the query string at runtime, and serialize it into a GraphQL-specific format

So they mean they didn't wanted to implement type check on their code side, and having the variables force it is they work around. That's nice and workable for languages where variable types can be very dynamic, so they are recommending this way for type precheck.
I feel that best practice is intended for web pages and things like postman, where the user can randomly add any input without anything forcing type before.
That's why I asked why should I add an extra variable layer and develop new logic to work around it to verify when I forward an integer to the code, there is really an integer, when python already made sure of it?
There are no documentation of what are our input variables and what is the input type to accommodate this best practice and I am not planning to deep dive code to figure that out. I don't think anyone would be comfortable creating the API calls in that way.
I needed to hunt for information for weeks just to make this basic version of it work.
In recommendation if those are the calls you are intending other's to make, then a documentation would be very helpful where all the input variables for each call is listed and described by what are the their input type, what they do, are they mandatory or not and what are the default ignore parameter, like "null". With examples for each how the call should look and not just a reference to check out the code and try to understand and untangle the information from there.

Reply:
> It does not remove the need to do the logic. I still need to cast Boolean in Python to string as GraphQL can't handle 'False', but only 'false', which is not a valid parameter for Boolean in Python. Same with 'null', Python has 'None', and GraphQL unable to handle that. So no, adding that overhead won't remove any type casting logic/code on Python side.

By utilizing GraphQL variables, it removes the need to typecast any of the variables datatypes on the query side. That's because the variables lookup is represented as a JSON object which Python can natively convert dict objects to. It's as simple as `json.dumps({'boolVal': False, 'stringVal': 'string', 'intVal': 0})`.

> There are no documentation of what are our input variables and what is the input type to accommodate this best practice and I am not planning to deep dive code to figure that out. I don't think anyone would be comfortable creating the API calls in that way.

GraphQL was designed to accommodate this type of design and we have documentation on the input types of every call in SWODLR's API build in as part of the API specification.

https://github.com/podaac/swodlr-api/tree/develop/src/main/resources/graphql

> I needed to hunt for information for weeks just to make this basic version of it work.

The input types are already defined as part of the API specification. The main work required here would be to parameterize the queries and to introduce a mechanism to input a variables dict to `GraphQlBase.Base_Post` like Frank suggested.

Reply:
> https://github.com/podaac/swodlr-api/tree/develop/src/main/resources/graphql
That is not a documentation, that is some code without any explanation... It does not make me understand what it should do, what were the mandatory and optional parameters, expected inputs, input ranges and so. I needed to ask you again and again for information.
Let me give an example of documentation:
https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/ecs/client/list_services.html

> The input types are already defined as part of the API specification. The main work required here would be to parameterize the queries and to introduce a mechanism to input a variables dict to GraphQlBase.Base_Post like Frank suggested.
And redesigning 70% of the tests as they are testing negative behavior that the "best practice" will mask or invalidate. As I mentioned before, I am not designing tests that will test best behavior, I am designing test that do edge cases and error handling, which is definitely not the domain of "best  practice".